x/ecocredit/server/marketplace/utils.go

```diff
@@ -18,3 +25,185 @@ func isDenomAllowed(ctx sdk.Context, denom string, pk ecocredit.ParamKeeper) boo
+	if err != nil {
+		if ormerrors.IsNotFound(err) {
+			buyerBal = &ecoApi.BatchBalance{
+				Address:  buyerAcc,
+				BatchId:  sellOrder.BatchId,
+				Tradable: "0",
+				Retired:  "0",
+				Escrowed: "0",
+			}
+		} else {
+			return err
+		}
+	}
+	if !opts.autoRetire { // if auto retire is disabled, we move the credits into the buyer's/supply's tradable balance.
+		tradableBalance, err := math.NewDecFromString(buyerBal.Tradable)
+		if err != nil {
+			return err
+		}
+		tradableBalance, err = math.SafeAddBalance(tradableBalance, purchaseQty)
+		if err != nil {
+			return err
+		}
+		buyerBal.Tradable = tradableBalance.String()
+
+		supplyTradable, err := math.NewDecFromString(supply.TradableAmount)
+		if err != nil {
+			return err
+		}
+		supplyTradable, err = math.SafeAddBalance(supplyTradable, purchaseQty)
+		if err != nil {
+			return err
+		}
+		supply.TradableAmount = supplyTradable.String()
+		if err = sdkCtx.EventManager().EmitTypedEvent(&core.EventReceive{
+			Sender:         sdk.AccAddress(sellOrder.Seller).String(),
+			Recipient:      buyerAcc.String(),
+			BatchDenom:     opts.batchDenom,
+			TradableAmount: purchaseQty.String(),
+			RetiredAmount:  "",
+			BasketDenom:    "",
+		}); err != nil {
+			return err
+		}
+	} else {
+		retiredBalance, err := math.NewDecFromString(buyerBal.Retired)
+		if err != nil {
+			return err
+		}
+		retiredBalance, err = math.SafeAddBalance(retiredBalance, purchaseQty)
+		if err != nil {
+			return err
+		}
+		buyerBal.Retired = retiredBalance.String()
+
+		supplyRetired, err := math.NewDecFromString(supply.RetiredAmount)
+		if err != nil {
+			return err
+		}
+		supplyRetired, err = math.SafeAddBalance(supplyRetired, purchaseQty)
+		if err != nil {
+			return err
+		}
+		supply.RetiredAmount = supplyRetired.String()
+		if err = sdkCtx.EventManager().EmitTypedEvent(&core.EventRetire{
+			Retirer:    buyerAcc.String(),
+			BatchDenom: opts.batchDenom,
+			Amount:     purchaseQty.String(),
+			Location:   opts.retirementLocation,
+		}); err != nil {
+			return err
+		}
+	}
+	if err = k.coreStore.BatchSupplyTable().Update(ctx, supply); err != nil {
+		return err
+	}
+	if err = k.coreStore.BatchBalanceTable().Save(ctx, buyerBal); err != nil {
+		return err
+	}
+
+	return k.bankKeeper.SendCoins(sdkCtx, buyerAcc, sellOrder.Seller, sdk.NewCoins(cost))
+}
+
+// getTotalCost calculates the cost of the order by multiplying the price per credit, and the amount of credits
+// desired in the order.
+func getTotalCost(pricePerCredit sdk.Int, amtCredits math.Dec) (sdk.Int, error) {
+	multiplier, err := math.NewPositiveFixedDecFromString(pricePerCredit.String(), amtCredits.NumDecimalPlaces())
+	if err != nil {
+		return sdk.Int{}, err
+	}
+	amountDec, err := amtCredits.Mul(multiplier)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	// divide by 1 to clear anything to the right of the decimal, so we can convert to sdk.Int
+	amountDec, err = amountDec.QuoInteger(math.NewDecFromInt64(1))
```

Comment:
please add TODO to update this code after https://github.com/regen-network/regen-ledger/issues/987

Reply:
added